servemux_test.go

```diff
@@ -596,6 +596,147 @@ func TestServeMux_MiddlewareSingle(t *testing.T) {
 	}
 }
 
+func TestServeMux_MiddlewareExceptFor(t *testing.T) {
+	s := NewServeMux()
+
+	s.MiddlewareExceptFor("/", mid1, http.MethodOptions, http.MethodPatch)
+	s.Handle("/", rightHandler)
+
+	req := httptest.NewRequest(http.MethodGet, "/", nil)
+
+	_, mids, _ := s.HandlerAndMiddleware(req)
+
+	if len(mids) != 1 {
+		t.Fatal("Wrong number of middlewares returned. Expected 1, got", len(mids))
+	}
+
+	req = httptest.NewRequest(http.MethodOptions, "/", nil)
+
+	_, mids, _ = s.HandlerAndMiddleware(req)
+
+	if len(mids) != 0 {
+		t.Fatal("Wrong number of middlewares returned. Expected 0, got", len(mids))
+	}
+
+	req = httptest.NewRequest(http.MethodPatch, "/", nil)
+
+	_, mids, _ = s.HandlerAndMiddleware(req)
+
+	if len(mids) != 0 {
+		t.Fatal("Wrong number of middlewares returned. Expected 0, got", len(mids))
+	}
+}
+
+func TestServeMux_MiddlewareExceptForOptions(t *testing.T) {
+	s := NewServeMux()
+
+	s.Route("/").MiddlewareExceptForOptions(mid1)
+	s.Handle("/", rightHandler)
+
+	req := httptest.NewRequest(http.MethodOptions, "/", nil)
+
+	_, mids, _ := s.HandlerAndMiddleware(req)
+
+	if len(mids) != 0 {
+		t.Fatal("Wrong number of middlewares returned. Expected 0, got", len(mids))
+	}
+}
+
+func TestServeMux_MiddlewareFor(t *testing.T) {
+	s := NewServeMux()
+
+	s.MiddlewareFor("/", mid1, http.MethodOptions, http.MethodPatch)
+	s.Handle("/", rightHandler)
+
+	req := httptest.NewRequest(http.MethodGet, "/", nil)
+
+	_, mids, _ := s.HandlerAndMiddleware(req)
+
+	if len(mids) != 0 {
+		t.Fatal("Wrong number of middlewares returned. Expected 0, got", len(mids))
+	}
+
+	req = httptest.NewRequest(http.MethodOptions, "/", nil)
+
+	_, mids, _ = s.HandlerAndMiddleware(req)
+
+	if len(mids) != 1 {
+		t.Fatal("Wrong number of middlewares returned. Expected 1, got", len(mids))
+	}
+
+	req = httptest.NewRequest(http.MethodPatch, "/", nil)
+
+	_, mids, _ = s.HandlerAndMiddleware(req)
+
+	if len(mids) != 1 {
+		t.Fatal("Wrong number of middlewares returned. Expected 1, got", len(mids))
+	}
+}
+
+func TestServeMux_MiddlewareFor_Panic(t *testing.T) {
+	s := NewServeMux()
+
+	defer func() {
+		err := recover()
+		if err == nil {
+			t.Error("Didn't panic")
+			return
+		}
+	}()
+
+	s.MiddlewareFor("/", mid1, http.MethodOptions, "llama")
+}
+
+func TestServeMux_MiddlewareFor_Nop(t *testing.T) {
+	s := NewServeMux()
+
+	s.MiddlewareFor("/", mid1)
+
+	req := httptest.NewRequest(http.MethodGet, "/", nil)
+
+	_, mids, _ := s.HandlerAndMiddleware(req)
+
+	if len(mids) != 0 {
+		t.Fatal("Wrong number of middlewares returned. Expected 0 got", len(mids))
+	}
+}
+
+func TestServeMux_MiddlewareExceptFor_Any(t *testing.T) {
+	s := NewServeMux()
+
+	s.MiddlewareExceptFor("/", mid1)
+
+	req := httptest.NewRequest(http.MethodGet, "/", nil)
```

Comment:
if we're not concerned about test time, we can loop through all methods for these "any" checks. The list of all methods can be kept as a constant for the file